Add archive bytes limit option to config

smarlowucf opened the pull request:
This sets the max length of the archive in bytes. If the archive grows larger than this limit the archive is trimmed until it satisfies the limit.

For k8s products this limit is 1MiB or 1048576 bytes. If the limit is set to 0 there is no size limit. The archive can grow as large as the retention period.

rtamalin reviewed (approved):
LGTM

rjschwei commented on csp_billing_adapter/archive.py:
```diff
@@ -18,24 +18,29 @@
 log = logging.getLogger('CSPBillingAdapter')
 
 DEFAULT_RETENTION_PERIOD = 6  # in months
+DEFAULT_BYTES_LIMIT = 0
```
I think the "special" unlimited size case should be a negative number, suggest `-1`. A size limit of `-1` would generally be nonsensical and as such bring attention to the setting and it's special meaning. This should also be considered in combination with the `DEFAULT_RETENTION_PERIOD` where 0 is not special and it simply means "no records are retained". Similar as `0` is certainly a valid limit for memory limit setting `0` would indicate "no records are retained" and `0` has the same meaning for both settings. Having the same interpretation for a numerical value for related settings simplifies the thought requirement, IMHO.

smarlowucf replied:
The smallest possible size of the archive is 2 bytes. Because a json string of an empty list is 2 bytes. So a value of 0 or 1 is non-sensical at least from a technical perspective. I can't think of examples off the top of my head but I'm certain I've used other tools where a value of 0 for a "limit" meant no limit. So I think there is precedence there. The affordances that I think we need are:

- Non-zero positive bytes limit
- Non-zero positive retention period limit
- No bytes limit (open discussion)
- Disable archive (retention period of zero)

I don't think we need two ways to disable the archive mechanism. If we want it then I would update the code to:

- Treat a 0 or 1 in bytes limit as "disable feature"
- Treat a 0 in retention period as "disable feature"
- Treat < 0 in bytes limit as no limit

This way min valid bytes limit is 2 which still makes no sense but prevents unnecessary handling to prevent infinite loop.

rjschwei replied:
Well actually from a memory perspective the smallest size is `53`
```
>>> sys.getsizeof(json.dumps('{}'))
53
```
and since `DEFAULT_BYTES_LIMIT` is supposed to set the memory usage limit for the retained records, driven by the k8s size limit of a config map we have to be concerned with the memory size.

It is fair to state that there is only 1 way to disable the feature and this is via the setting of the `DEFAULT_RETENTION_PERIOD`, then let's have a comment to that effect in the code.

My concern is not about the precedence that `0` has the special "no limit" meaning. My concern is that the reader of the code has 2 assignments in succession for which the value `0` has two distinct meanings with no explanation and where `0` is, based on naming of the variables, a valid setting that does not imply a special meaning. The question is how do we, in the code, at the point of assignment, make the special meaning evident? This can be done by using a value that on the surface makes not sense, `-1` for example, or it can be done via comment at the point assignment.

Ultimately we do not want to create a cognitive burden for the code reader that thinks of `0` as nothing special at this point and then ~90 lines later has to learn it is special after all.

smarlowucf replied:
The measurement that matters is the size per character not size of the python object memory footprint.

rjschwei commented on csp_billing_adapter/archive.py:
```diff
@@ -44,17 +49,34 @@ def append_metering_records(
+        # empty list is 2 bytes.
+        while True:
+            # Trim archive until it is smaller than max bytes
+            archive_size = len(bytes(json.dumps(archive), 'utf-8'))
```
The calculation is not correct, the `bytes()` function turns the string that `json.dumps()` produces into a list of integers, an int for every character. Then `len()` gives us the number of integers in the list. However, this is not equal to the size of the data structure from a memory footprint perspective. An integer object in Python consumes 28 bytes
```
>>> import sys
>>> sys.getsizeof(1)
28
```
Meaning the calculated `archive_size` needs to be multiplied by `28` for the comparison to `max_bytes` since we are concerned to not exceed the memory limit of a config map in k8s.

smarlowucf replied:
No it's correct:

```
import json


archive = [
    {
        "bill_time": "2024-01-03T20:06:42.076972+00:00",
        "metering_id": "FFFFFFFF-FFFF-FFFF-FFFF-FFFFFFFFFFFF",
        "usage_records": [
            {
                "managed_node_count": 9,
                "managed_systems": 6,
                "reporting_time": "2024-01-03T20:06:42.076972+00:00"
            }
        ]
    }
]

archive_size = len(bytes(json.dumps(archive), 'utf-8'))

with open('/tmp/t.json', 'w') as f:
    json.dump(archive, f)

print(f'computed size: {archive_size}')
```

```
16:37:02 ▶ python3 /tmp/t.py
computed size: 228
16:37:05 ▶ l /tmp/t.json
-rw-r--r-- 1 smarlow smarlow 228 Jan 10 16:37 /tmp/t.json
```

smarlowucf replied:
I guess there's no need to convert to bytes though. It should only contain ascii characters which are 1 byte per character.

smarlowucf replied:
getsizeof is returning the size of the python object not the size of the character that will be stored.

smarlowucf commented:
Made changes per comments:

- Dropped the default values as these are none functional. A value has to be provided in config anyway

Changed how the values are treated to the following:

- Treat a 0 or 1 in bytes limit as "disable feature"
- Treat a 0 in retention period as "disable feature"
- Treat < 0 in bytes limit as no limit
